wgpu-hal/src/dynamic/mod.rs

```diff
@@ -0,0 +1,67 @@
+use std::any::Any;
+
+use wgt::WasmNotSendSync;
+
+/// Base trait for all resources, allows downcasting via `std::any::Any`.
+pub trait DynResource: Any + WasmNotSendSync + 'static {
+    fn as_any(&self) -> &dyn Any;
+    fn as_any_mut(&mut self) -> &mut dyn Any;
+}
+
+/// Utility macro for implementing `DynResource` for a list of types.
+macro_rules! impl_dyn_resource {
+    ($($type:ty),*) => {
+        $(
+            impl crate::DynResource for $type {
+                fn as_any(&self) -> &dyn std::any::Any {
+                    self
+                }
+
+                fn as_any_mut(&mut self) -> &mut dyn std::any::Any {
+                    self
+                }
+            }
+        )*
+    };
+}
+pub(crate) use impl_dyn_resource;
+
+/// Extension trait for `DynResource` used by by implementations of various dynamic resource traits.
+trait DynResourceExt {
+    fn expect_downcast_ref<T: DynResource>(&self) -> &T;
+    fn expect_downcast_mut<T: DynResource>(&mut self) -> &mut T;
```

Comment:
yeah, that was the idea. I wanted to encapsulate the `expect` calls because this falls under the "HAL has many assumptions and is inherently unsafe, you have to use it correctly" umbrella